librad/src/net/upgrade.rs

```diff
@@ -19,36 +19,60 @@
 //! (irrespective of ALPN, which applies per-connection). This module implements
 //! the negotiation protocol.
 
-use std::{io, marker::PhantomData, ops::Deref, pin::Pin};
+use std::{
+    fmt::{self, Debug, Display},
+    io,
+    marker::PhantomData,
+    ops::Deref,
+    pin::Pin,
+    time::Duration,
+};
 
 use futures::{
-    future::{BoxFuture, FutureExt},
-    io::{AsyncRead, AsyncWrite},
-    sink::SinkExt,
-    stream::TryStreamExt,
+    future::{self, TryFutureExt as _},
+    io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt},
     task::{Context, Poll},
 };
-use futures_codec::Framed;
-use minicbor::{Decode, Encode};
+use futures_timer::Delay;
 use thiserror::Error;
 
-use crate::{
-    git::p2p::transport::GitStream,
-    net::{
-        codec::{CborCodec, CborCodecError},
-        quic,
-    },
-};
+use crate::git::p2p::transport::GitStream;
+
+/// Timeout waiting for an [`UpgradeRequest`].
+const RECV_UPGRADE_TIMEOUT: Duration = Duration::from_secs(23);
+
+/// Length in bytes of the CBOR encoding of [`UpgradeRequest`].
+///
+/// We use this to allocate only a fixed-size buffer, and not deal with
+/// unconsumed bytes. Make sure to adjust in case [`UpgradeRequest`] gains
+/// larger variants.
+const UPGRADE_REQUEST_ENCODING_LEN: usize = 4;
```

Comment:
Would the tests break if a new variant was added?

I'm wondering how we can catch this early if someone added a variant but didn't update this constant.

Reply:
Yes, `test_upgrade` will break if 4 bytes is not enough